sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -148,80 +158,111 @@ def fit(self, X, y):
+                    # evaluating the training loss, and create
+                    # raw_predictions_val for storing the raw predictions of
+                    # the validation data.
+
+                    if self._use_validation_data:
+                        raw_predictions_val = np.zeros(
+                            shape=(self.n_trees_per_iteration_,
+                                   X_binned_val.shape[0]),
+                            dtype=self._baseline_prediction.dtype
+                        )
+
+                        raw_predictions_val += self._baseline_prediction
+
+                    self._check_early_stopping_loss(raw_predictions, y_train,
+                                                    raw_predictions_val, y_val)
+                else:
+                    self.scorer_ = check_scoring(self, self.scoring)
+                    # scorer_ is a callable with signature (est, X, y) and
+                    # calls est.predict() or est.predict_proba() depending on
+                    # its nature.
+                    # Unfortunately, each call to scorer_() will compute
+                    # the predictions of all the trees. So we use a subset of
+                    # the training set to compute train scores.
+                    subsample_size = 10000  # should we expose this parameter?
+                    indices = np.arange(X_binned_train.shape[0])
+                    self._indices = indices
+                    if X_binned_train.shape[0] > subsample_size:
+                        # TODO: not critical but stratify using resample()
+                        indices = rng.choice(indices, subsample_size,
+                                             replace=False)
+                    X_binned_small_train = X_binned_train[indices]
+                    y_small_train = y_train[indices]
+                    # Predicting is faster on C-contiguous arrays.
+                    X_binned_small_train = np.ascontiguousarray(
+                        X_binned_small_train)
+
+                    self._check_early_stopping_scorer(
+                        X_binned_small_train, y_small_train,
+                        X_binned_val, y_val,
+                    )
+            begin_at_stage = 0
+
+        else:
+            # Convert array attributes to lists
+            self._tolist_state()
+
+            # Compute raw predictions
+            raw_predictions = self._raw_predict(X_binned_train)
+            if hasattr(self, '_indices'):
+                X_binned_small_train = X_binned_train[self._indices]
+                y_small_train = y_train[self._indices]
                 X_binned_small_train = np.ascontiguousarray(
                     X_binned_small_train)
 
-                self._check_early_stopping_scorer(
-                    X_binned_small_train, y_small_train,
-                    X_binned_val, y_val,
-                )
+            # Get the gradients and hessians from the previous fit
```

Comment:
You shouldn't need this: gradients and hessians are computed below in `update_gradient_and_hessians()`, according to the `raw_predictions` that you just updated.

Reply:
`update_gradients_and_hessians()` requires `gradients` and `hessians` as parameters. I can initialize them with:
```python
gradients, hessians = self.loss_.init_gradients_and_hessians(
    n_samples=n_samples,
    prediction_dim=self.n_trees_per_iteration_
 )
```
but I am not sure that it will give the expected results.

IMO:
* either `update_gradients_and_hessians()` should not require `gradients` and `hessians` and they can be initialized,
* or `update_gradients_and_hessians()` requires `gradients` and `hessians` and the last values of `gradients` and `hessians` should be saved.

Reply:
Maybe `update` is just a bad name for it, but `update_gradients_and_hessians` does not use the previous values of the gradients and hessians. It just compute, and modifies the arrays inplace (that's why we pass them, so that we don't have to allocate them at each iteration

calling `loss_.init_gradients_and_hessians` should be enough

Reply:
I did not think about that optimization (I must say that it is beyond my skillset right now), so thanks for the explanation and I'll try to think about it next time :)